metrics/sliding_time_window_sample.go

```diff
@@ -0,0 +1,191 @@
+package metrics
+
+// go version of sliding window reservoir
+// refer to : https://github.com/dropwizard/metrics/blob/3.2-development/metrics-core/src/main/java/com/codahale/metrics/SlidingTimeWindowReservoir.java
+// and : https://github.com/rcrowley/go-metrics/blob/master/sample.go
+import (
+	"sync"
+	"time"
+
+	gomet "github.com/rcrowley/go-metrics"
+)
+
+const (
+	collisionBuffer int64 = 256
+	trimThreshold   int64 = 256
+	clearBuffer     int64 = int64(time.Hour/time.Nanosecond) * 256
+)
+
+// SlidingTimeWindowSample : http://metrics.dropwizard.io/3.1.0/manual/core/#sliding--time-window-reservoirs
+type SlidingTimeWindowSample struct {
+	window   int64
+	lastTick int64
+	count    int64
+	mutex    sync.Mutex
+	values   map[int64]int64
+}
+
+// NewSlidingTimeWindowSample : New
+func NewSlidingTimeWindowSample(duration string) gomet.Sample {
+	if gomet.UseNilMetrics {
+		return gomet.NilSample{}
+	}
+	sample := &SlidingTimeWindowSample{
+		values:   make(map[int64]int64),
+		lastTick: time.Now().UnixNano() * collisionBuffer,
+	}
+	if dur, err := time.ParseDuration(duration); err == nil {
+		sample.window = dur.Nanoseconds() * collisionBuffer
+	}
+	return sample
+}
+
+func (s *SlidingTimeWindowSample) trim() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	now := s.getTick()
+	windowStart := now - s.window
+	windowEnd := now + clearBuffer
+	if windowStart < windowEnd {
+		for k := range s.values {
+			if k < windowStart {
+				delete(s.values, k)
+			}
+			if k >= windowEnd {
+				delete(s.values, k)
+			}
+		}
+	} else {
+		for k := range s.values {
+			if windowEnd < k && k < windowStart {
+				delete(s.values, k)
+			}
+		}
+	}
+}
+
+func (s *SlidingTimeWindowSample) getTick() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	oldTick := s.lastTick
+	tick := time.Now().UnixNano()
+	if tick-oldTick > 0 {
+		s.lastTick = tick
+	} else {
+		s.lastTick = oldTick + 1
+	}
+	return s.lastTick
+}
+
+// Count : implements interface gomet.Sample
+func (s *SlidingTimeWindowSample) Count() int64 {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.count
+}
+
+// Max : implements interface gomet.Sample
+func (s *SlidingTimeWindowSample) Max() int64 {
+	s.mutex.Lock()
```

Comment:
I'm implement a gomet.Sample interface to create a new Sampler. so i've to write thread-safe code myself.
reference: [here](https://github.com/rcrowley/go-metrics/blob/master/sample.go)